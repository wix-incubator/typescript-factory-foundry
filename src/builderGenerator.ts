import {
  ClassDeclarationStructure,
  FunctionDeclarationStructure,
  ImportDeclarationStructure,
  IndentationText,
  InterfaceDeclaration,
  MethodDeclarationStructure,
  Project,
  QuoteKind,
  Scope,
  StructureKind,
  ts,
  Type,
  TypeAliasDeclaration,
} from 'ts-morph';
import * as path from 'path';
import { copyFile } from 'fs/promises';

export async function generateBuilders(
  inputFilePath: string,
  outputFilePath: string,
  { enableSubBuilders = false } = {},
): Promise<void> {
  const project = new Project({
    manipulationSettings: {
      indentationText: IndentationText.TwoSpaces,
      quoteKind: QuoteKind.Single,
    },
    compilerOptions: {
      strictNullChecks: true,
    },
  });
  project.addSourceFileAtPath(inputFilePath);
  const fileName = path.basename(inputFilePath);
  const fileNameWithoutExtension = path.basename(
    inputFilePath,
    path.extname(inputFilePath),
  );
  const sourceFile = project.getSourceFileOrThrow(fileName);

  project.createDirectory(outputFilePath);

  const allTypes = sourceFile.getTypeAliases();
  const allInterfaces = sourceFile.getInterfaces();

  const classesAndFactories: {
    classObj: ClassDeclarationStructure;
    builderFunc: FunctionDeclarationStructure;
    additionalImports: ImportDeclarationStructure[];
  }[] = [];

  const indexFile = project.createSourceFile(
    path.join(outputFilePath, 'index.ts'),
    undefined,
    {
      overwrite: true,
      scriptKind: ts.ScriptKind.TS,
    },
  );

  for (const typeEntry of [...allTypes, ...allInterfaces]) {
    const { classObj, builderFunc, additionalImports } = generateBuilderFunc(
      typeEntry,
      {
        enableSubBuilders,
      },
    );
    if (classObj && builderFunc) {
      classesAndFactories.push({
        classObj,
        builderFunc,
        additionalImports,
      });
    }
  }

  const promises = classesAndFactories.map(
    ({ classObj, builderFunc, additionalImports }) => {
      const outFile = project.createSourceFile(
        path.join(outputFilePath, `${classObj.name}.ts`),
        undefined,
        {
          overwrite: true,
          scriptKind: ts.ScriptKind.TS,
        },
      );

      outFile.addStatements([
        '/* eslint-disable */',
        {
          kind: StructureKind.ImportDeclaration,
          namedImports: ['DeepPartial'],
          moduleSpecifier: 'ts-essentials',
        },
        {
          kind: StructureKind.ImportDeclaration,
          namespaceImport: 'SchemaTypes',
          moduleSpecifier: `./${fileNameWithoutExtension}`,
        },
        ...additionalImports,
      ]);

      indexFile.addExportDeclaration({
        kind: StructureKind.ExportDeclaration,
        moduleSpecifier: `./${classObj.name}`,
      });

      outFile.addClass(classObj);
      outFile.addFunction(builderFunc);

      return outFile
        .save()
        .then(() => copyFile(inputFilePath, `${outputFilePath}/${fileName}`));
    },
  );

  return Promise.all([...promises, indexFile.save()]).then();
}

function generateBuilderFunc(
  typeAlias: TypeAliasDeclaration | InterfaceDeclaration,
  { enableSubBuilders }: { enableSubBuilders: boolean },
) {
  const rootType = typeAlias.getType();
  const additionalImports: ImportDeclarationStructure[] = [];

  const props = rootType.getProperties();

  if (props.length <= 0) {
    return {};
  }

  if (rootType.isUnion()) {
    console.warn(`Skipped type generation for union: ${typeAlias.getName()}`);

    return {};
  }

  const objName = 'obj';
  const typeName = typeAlias.getName();
  const className = `${typeName}Builder`;

  const methods: MethodDeclarationStructure[] = [];
  for (const prop of props) {
    const propName = prop.getName();
    if (propName === '__typename') {
      continue;
    }
    const type = prop.getTypeAtLocation(prop.getValueDeclarationOrThrow());

    const unionTypes = type.isUnion() ? type.getUnionTypes() : [type];

    const parametersType = unionTypes
      .map(builderParameterTypeMapper)
      .filter((t) => t !== 'false')
      .map((t) => (t === 'true' ? 'boolean' : t))
      .join(' | ');

    const subBuilderName =
      !isPrimitive(type) && !type.isArray() && !type.getText().startsWith('{')
        ? `${type.getText().replace(/import\("[^"]*"\)\./gi, '')}Builder`
        : null;

    if (enableSubBuilders && subBuilderName) {
      additionalImports.push({
        kind: StructureKind.ImportDeclaration,
        namedImports: [subBuilderName],
        moduleSpecifier: `./${subBuilderName}`,
      });
    }

    const method: MethodDeclarationStructure = {
      kind: StructureKind.Method,
      name: `with${capitalizeFirstLetter(propName)}`,
      returnType: className,
      typeParameters: [`P extends ${parametersType}`],
      parameters: [
        {
          name: 'val',
          type: `P${
            enableSubBuilders && subBuilderName
              ? ` | ((b: ${subBuilderName}) => ${subBuilderName})`
              : ''
          }`,
        },
      ],
      scope: Scope.Public,
      statements: [
        ...(enableSubBuilders && subBuilderName
          ? [
              `if (typeof val === 'function') {`,
              `  this.obj.${propName} = val(new ${subBuilderName}(this.obj.${propName})).get();`,
              '  return this;',
              '}',
            ]
          : []),
        `this.obj.${propName} = val;`,
        'return this;',
      ],
    };
    methods.push(method);
  }

  methods.push({
    kind: StructureKind.Method,
    name: 'includeTypename',
    scope: Scope.Public,
    statements: [
      '// @ts-ignore',
      `this.obj.__typename = '${typeName}';`,
      'return this;',
    ],
  });

  methods.push({
    kind: StructureKind.Method,
    name: 'get',
    scope: Scope.Public,
    returnType: `SchemaTypes.${typeName}`,
    statements: `return Object.assign({}, this.${objName}) as SchemaTypes.${typeName};`,
  });

  const builderFunc: FunctionDeclarationStructure = {
    kind: StructureKind.Function,
    name: `a${className}`,
    isExported: true,
    typeParameters: [`O extends DeepPartial<SchemaTypes.${typeName}> = {}`],
    returnType: `${className}`,
    parameters: [
      {
        name: 'baseObj',
        type: 'O',
        hasQuestionToken: true,
      },
    ],
    statements: `return new ${className}(baseObj ?? {});`,
  };

  const classObj: ClassDeclarationStructure = {
    kind: StructureKind.Class,
    name: className,
    ctors: [
      {
        kind: StructureKind.Constructor,
        parameters: [
          {
            name: objName,
            initializer: `{} as DeepPartial<SchemaTypes.${typeName}>`,
            isReadonly: true,
            scope: Scope.Private,
          },
        ],
      },
    ],
    methods,
    isExported: true,
  };

  return { builderFunc, classObj, additionalImports };
}

function builderParameterTypeMapper(subType: Type): string {
  const isArray = subType.isArray();

  const resultType = isArray ? subType.getArrayElementType()! : subType;

  let resultTypeStr = resultType
    .getText()
    .replace(/import\("[^"]*"\)/gi, 'SchemaTypes');

  resultTypeStr = isPrimitive(resultType)
    ? `${resultTypeStr}`
    : `DeepPartial<${resultTypeStr}>`;

  return isArray ? `${resultTypeStr}[]` : resultTypeStr;
}

function isPrimitive(type: Type) {
  return [
    type.isString(),
    type.isNumber(),
    type.isLiteral(),
    type.isBoolean(),
    type.isUndefined(),
    type.isNull(),
    type.isAny(),
    type.isEnum(),
  ].some(Boolean);
}

function capitalizeFirstLetter(string: string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}
