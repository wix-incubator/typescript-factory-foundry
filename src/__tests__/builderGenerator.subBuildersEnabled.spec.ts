import * as path from 'path';
import { generateBuilders } from '../builderGenerator';
import * as fs from 'fs';

jest.setTimeout(15_000);

describe('Builder Generator { enableSubBuilders: true }', () => {
  it('should generate builders for interfaces and types', async () => {
    const simpleBuildersExample = path.join(
      __dirname,
      './examples/simpleInterfacesAndTypes.ts',
    );
    const outputFolder = path.join(
      __dirname,
      'output/simple-interfaces-and-types-enabled-nested',
    );
    cleanOutput(outputFolder);

    await generateBuilders(simpleBuildersExample, outputFolder, {
      enableSubBuilders: true,
    });

    const indexFile = fs.readFileSync(`${outputFolder}/index.ts`, 'utf8');
    const ScalarsBuilderFile = fs.readFileSync(
      `${outputFolder}/ScalarsBuilder.ts`,
      'utf8',
    );
    const SimpleExampleBuilderFile = fs.readFileSync(
      `${outputFolder}/SimpleExampleBuilder.ts`,
      'utf8',
    );
    const simpleInterfacesAndTypesFile = fs.readFileSync(
      `${outputFolder}/simpleInterfacesAndTypes.ts`,
      'utf8',
    );
    const NestedTypeExampleBuilderFile = fs.readFileSync(
      `${outputFolder}/NestedTypeExampleBuilder.ts`,
      'utf8',
    );

    expect(fs.readdirSync(outputFolder)).toEqual([
      'NestedTypeExampleBuilder.ts',
      'ScalarsBuilder.ts',
      'SimpleExampleBuilder.ts',
      'index.ts',
      'simpleInterfacesAndTypes.ts',
    ]);
    expect(NestedTypeExampleBuilderFile).toMatchSnapshot();
    expect(ScalarsBuilderFile).toMatchSnapshot();
    expect(SimpleExampleBuilderFile).toMatchSnapshot();
    expect(simpleInterfacesAndTypesFile).toMatchSnapshot();
    expect(indexFile).toMatchSnapshot();
  });
});

function cleanOutput(outputFolder: string) {
  if (fs.existsSync(outputFolder)) {
    fs.rmSync(outputFolder, { recursive: true, force: true });
  }
}
