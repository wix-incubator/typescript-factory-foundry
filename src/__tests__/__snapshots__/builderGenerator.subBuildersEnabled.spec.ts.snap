// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder Generator { enableSubBuilders: true } should generate builders for interfaces and types 1`] = `
"/* eslint-disable */
import { DeepPartial } from 'ts-essentials';
import * as SchemaTypes from './simpleInterfacesAndTypes';
import { SimpleExampleBuilder } from './SimpleExampleBuilder';
import { ScalarsBuilder } from './ScalarsBuilder';

export class NestedTypeExampleBuilder {
  constructor(private readonly obj = {} as DeepPartial<SchemaTypes.NestedTypeExample>) {
  }

  public withNestedProp1<P extends DeepPartial<SchemaTypes.SimpleExample>>(val: P | ((b: SimpleExampleBuilder) => SimpleExampleBuilder)): NestedTypeExampleBuilder {
    if (typeof val === 'function') {
      this.obj.nestedProp1 = val(new SimpleExampleBuilder(this.obj.nestedProp1)).get();
      return this;
    }

    this.obj.nestedProp1 = val;
    return this;
  }

  public withNestedProp2<P extends DeepPartial<{ simple: SchemaTypes.SimpleExample; }>>(val: P): NestedTypeExampleBuilder {
    this.obj.nestedProp2 = val;
    return this;
  }

  public withScalar1<P extends DeepPartial<SchemaTypes.Scalars>>(val: P | ((b: ScalarsBuilder) => ScalarsBuilder)): NestedTypeExampleBuilder {
    if (typeof val === 'function') {
      this.obj.scalar1 = val(new ScalarsBuilder(this.obj.scalar1)).get();
      return this;
    }

    this.obj.scalar1 = val;
    return this;
  }

  public includeTypename() {
    // @ts-ignore
    this.obj.__typename = 'NestedTypeExample';
    return this;
  }

  public get(): SchemaTypes.NestedTypeExample {
    return Object.assign({}, this.obj) as SchemaTypes.NestedTypeExample;
  }
}

export function aNestedTypeExampleBuilder<O extends DeepPartial<SchemaTypes.NestedTypeExample> = {}>(baseObj?: O): NestedTypeExampleBuilder {
  return new NestedTypeExampleBuilder(baseObj ?? {});
}
"
`;

exports[`Builder Generator { enableSubBuilders: true } should generate builders for interfaces and types 2`] = `
"/* eslint-disable */
import { DeepPartial } from 'ts-essentials';
import * as SchemaTypes from './simpleInterfacesAndTypes';

export class ScalarsBuilder {
  constructor(private readonly obj = {} as DeepPartial<SchemaTypes.Scalars>) {
  }

  public withID<P extends string>(val: P): ScalarsBuilder {
    this.obj.ID = val;
    return this;
  }

  public withString<P extends string>(val: P): ScalarsBuilder {
    this.obj.String = val;
    return this;
  }

  public withBoolean<P extends boolean>(val: P): ScalarsBuilder {
    this.obj.Boolean = val;
    return this;
  }

  public withInt<P extends number>(val: P): ScalarsBuilder {
    this.obj.Int = val;
    return this;
  }

  public withFloat<P extends number>(val: P): ScalarsBuilder {
    this.obj.Float = val;
    return this;
  }

  public withBigDecimal<P extends number>(val: P): ScalarsBuilder {
    this.obj.BigDecimal = val;
    return this;
  }

  public withJson<P extends any>(val: P): ScalarsBuilder {
    this.obj.Json = val;
    return this;
  }

  public withLong<P extends number>(val: P): ScalarsBuilder {
    this.obj.Long = val;
    return this;
  }

  public includeTypename() {
    // @ts-ignore
    this.obj.__typename = 'Scalars';
    return this;
  }

  public get(): SchemaTypes.Scalars {
    return Object.assign({}, this.obj) as SchemaTypes.Scalars;
  }
}

export function aScalarsBuilder<O extends DeepPartial<SchemaTypes.Scalars> = {}>(baseObj?: O): ScalarsBuilder {
  return new ScalarsBuilder(baseObj ?? {});
}
"
`;

exports[`Builder Generator { enableSubBuilders: true } should generate builders for interfaces and types 3`] = `
"/* eslint-disable */
import { DeepPartial } from 'ts-essentials';
import * as SchemaTypes from './simpleInterfacesAndTypes';

export class SimpleExampleBuilder {
  constructor(private readonly obj = {} as DeepPartial<SchemaTypes.SimpleExample>) {
  }

  public withProp1<P extends boolean>(val: P): SimpleExampleBuilder {
    this.obj.prop1 = val;
    return this;
  }

  public withProp2<P extends number>(val: P): SimpleExampleBuilder {
    this.obj.prop2 = val;
    return this;
  }

  public withProp3<P extends DeepPartial<{ innerProp: string; }>>(val: P): SimpleExampleBuilder {
    this.obj.prop3 = val;
    return this;
  }

  public includeTypename() {
    // @ts-ignore
    this.obj.__typename = 'SimpleExample';
    return this;
  }

  public get(): SchemaTypes.SimpleExample {
    return Object.assign({}, this.obj) as SchemaTypes.SimpleExample;
  }
}

export function aSimpleExampleBuilder<O extends DeepPartial<SchemaTypes.SimpleExample> = {}>(baseObj?: O): SimpleExampleBuilder {
  return new SimpleExampleBuilder(baseObj ?? {});
}
"
`;

exports[`Builder Generator { enableSubBuilders: true } should generate builders for interfaces and types 4`] = `
"export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  BigDecimal: number;
  Json: any;
  Long: number;
};

export interface SimpleExample {
  prop1: Scalars['Boolean'];
  prop2: number;
  prop3: {
    innerProp: string;
  };
}

export type NestedTypeExample = {
  nestedProp1: SimpleExample;
  nestedProp2: {
    simple: SimpleExample;
  };
  scalar1: Scalars;
};
"
`;

exports[`Builder Generator { enableSubBuilders: true } should generate builders for interfaces and types 5`] = `
"export * from './ScalarsBuilder';
export * from './NestedTypeExampleBuilder';
export * from './SimpleExampleBuilder';
"
`;
